src/libsyntax/parse/lexer/mod.rs

```diff
@@ -121,6 +132,7 @@ impl<'a> StringReader<'a> {
             sp: self.peek_span,
         };
         self.advance_token()?;
+        self.span_src_raw = self.peek_span_src_raw;
```

Comment:
Heh, I've just spend like an hour perusing this line, and I came to the conclusion that:

* this code is buggy
* the code is observationaly correct

The code is buggy because, while `span` and `token` point to *this* token, `span_src` points to the *next* token. In other words, `span_src_raw` and `peek_span_src_raw` are just equal all the time, because, first `advance_token` sets `peek_span_src_raw`, and then we assign `peek_span_src_raw` to `span_src_raw`.

However, this bug does not show up, because we only use peek span to check for jointness. And if spans A and B are adjacent, than spans *following* A and B are adjacent as well.

In other words, for `++ ` we don't check that first `+` is adjacent to the next, we check that the token following the first `+` (the second `+`) is adjacent to the token, following the second `+` (whitespace).

Similarly, for `+ + `  we actually check that *whitespace* following `+` is not adjacent.